[stdlib] Adopt conditional conformance for Indices, Slice, ReversedCollection

airspeedswift opened the pull request:
This PR refactors the variations of `Slice`, `DefaultIndices`, and `ReversedCollection` to be conditional conformances on each base type e.g.

```swift
ReversedRandomAccessCollection<Base: RandomAccessCollection>: RandomAccessCollection
```

becomes:

```swift
extension ReversedCollection: RandomAccessCollection where Base: RandomAccessCollection { }

@available(*, deprecated, renamed: "ReversedCollection")
public typealias ReversedRandomAccessCollection<T: RandomAccessCollection> = ReversedCollection<T>
```

airspeedswift committed:
Replace ReversedRandomAccessCollection with a conditional extension

airspeedswift committed:
Refactor Indices and Slice to use conditional conformance

Refactor into base definitions + extensions.

Redundant associated types are still required to word around bugs in associated type inference.

slavapestov commented:
Wow that's a lot of deleted code.

moiseev commented on stdlib/public/core/Indices.swift:
```diff
@@ -135,15 +119,16 @@ where Indices == ${Self}<Self> {
 }
 
-% end
+public typealias DefaultBidirectionalIndices<T: BidirectionalCollection> = DefaultIndices<T>
```
Deprecated in swift 4.1, perhaps?

moiseev commented on stdlib/public/core/RangeReplaceableCollection.swift:
```diff
@@ -1,4 +1,4 @@
-//===--- RangeReplaceableCollection.swift.gyb -----------------*- swift -*-===//
+//===--- RangeReplaceableCollection.swift -----------------*- swift -*-===//
```
Notpick: No longer 80 columns.

moiseev commented on stdlib/public/core/Reverse.swift:
```diff
@@ -269,177 +269,11 @@ public struct ReversedCollection<
-  public let _base: Base
-}
+@available(*, deprecated, renamed: "ReversedCollection")
+typealias ReversedRandomAccessCollection<T> = ReversedCollection<T> where T: RandomAccessCollection
```
`public`

natecook1000 commented on stdlib/public/core/Indices.swift:
```diff
@@ -115,8 +84,23 @@ public struct ${Self}<
+  }
+}
+
+extension DefaultIndices: RandomAccessCollection where Elements: RandomAccessCollection { }
```
Need to make sure we're forwarding distance(to:) and index(_:offsetBy:) to the underlying collection.

natecook1000 commented on stdlib/public/core/Reverse.swift:
```diff
@@ -269,177 +269,11 @@ public struct ReversedCollection<
-  public var endIndex: Index {
-    return ReversedRandomAccessIndex(_base.startIndex)
-  }
+extension ReversedCollection: RandomAccessCollection where Base: RandomAccessCollection { }
```
Here too — need the index-moving methods.

natecook1000 reviewed (commented):
This is so great! :clap::clap::clap:

DougGregor commented:
The first compiler crash this hits is: https://bugs.swift.org/browse/SR-6466

DougGregor commented on stdlib/public/core/Reverse.swift:
```diff
@@ -523,22 +314,6 @@ extension LazyCollectionProtocol
-  ///
-  /// - Complexity: O(1)
-  @_inlineable
-  public func reversed() -> LazyRandomAccessCollection<
```
Whoops, we still need this, up until the point where the Lazy collections also use conditional conformances

DougGregor committed:
Fix test cases for adoption of conditional conformances.

These fixes are for diagnostics that changed in benign ways or for
places in the testsuite that were taking advantage of standard library
internals.

DougGregor committed:
[RangeReplaceableCollection] Eliminate unnecessary slicing subscript operator.

moiseev commented:
@swift-ci Please smoke test macOS platrform

moiseev commented:
@swift-ci Please smoke test macOS platform

DougGregor commented:
@airspeedswift , that last commit re-breaks the test. The gyb'd conformances *are* redundant, so we (correctly) "redundant conformance" errors. Why did you revert my change? Because it uncovered a source-compatibility issue?

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
Quite curious how this fares on the compatibility suite in its current state.

DougGregor commented:
Note that *all* of the test suite failure other that the `Prototypes/Algorithms.swift.gyb` one mentioned about are in the `subscript(_: Range)/Set/semantics` test at https://github.com/apple/swift/blob/master/stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb#L244-L295.

DougGregor commented:
... and the crash appears to be in the `materializeForSet` for Slice's range subscript.

```
_T0s5SliceVss17MutableCollectionRzrlEAByxGs5RangeVy5IndexQzGcimytfU_ ---> closure #1 : () in (extension in Swift):Swift.Slice< where A: Swift.MutableCollection>.subscript.materializeForSet : (Swift.Range<A.Index>) -> Swift.Slice<A>
```

DougGregor commented:
The Plank and R failures are the SwiftPM failure under discussion in https://github.com/apple/swift-package-manager/pull/1393, so this isn't introducing any source-compatibility regressions.

DougGregor commented:
Turns out that we don't need that big pile of explicit typealiases in the standard library now that I've reinstated the associated type redeclarations in the `Collection` lattice.

airspeedswift commented:
@DougGregor sorry, I thought that was reverting a change I made from before you restricted the feature to only stdlib use (the test was failing). Yeah, if that original code doesn't compile, it does suggest a source compatibility issue we'll need to work around.

moiseev commented:
@swift-ci Please smoke test macOS platform

DougGregor commented:
The source compatibility issue here is that the code in question introduced conformances for each of the `Revered*Collection` types to a single new protocol, e.g.,

```swift
protocol P { }
extension ReversedCollection : P { }
extension ReversedRandomAccessCollection : P { }
extension ReversedRangeReplaceableCollection : P { }
```

With this PR, those conformances are now redundant, because it's stating `ReversedCollection : P ` three times.

airspeedswift commented:
@swift-ci please smoke test compiler performance

moiseev commented:
@swift-ci Please test macOS platform

moiseev commented:
@swift-ci please smoke test compiler performance

moiseev commented:
@swift-ci Please smoke benchmark

DougGregor commented:
FWIW, I rebased on master, fixed up the conflicts, and removed the extra hack + workaround commits.

DougGregor committed:
Add -enable-experimental-conditional-conformances to test.

DougGregor commented:
Fixed up the `Prototypes/Algorithms.swift.gyb test`. The "only" remaining failures are those due to https://github.com/apple/swift/blob/master/stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb#L244-L295.

airspeedswift committed:
Gruesome workaround for crasher in MutableSlice tests

airspeedswift commented:
@swift-ci please test

airspeedswift commented:
The only significant casualty on the benchmarks seems to be CharIteration in Debug. @milseman possibly something to be aware of...

airspeedswift commented:
@swift-ci please test source compatibility

airspeedswift commented:
@swift-ci please test compiler performance

airspeedswift commented:
@swift-ci please test

DougGregor commented:
@airspeedswift merge it! merge it! :)

natecook1000 commented:
Did we figure out a plan for https://github.com/apple/swift/pull/12913#issuecomment-347975121 ? If that's part of our compatibility suite it'll surely be elsewhere in the wild—I could definitely see people giving conformance to a new protocol to multiple `Slice` types, for instance. Do we need to keep these types around in deprecated form to have compatibility with Swift 4.0?

DougGregor commented:
@natecook1000, that source-compatibility issue was in our own standard library test suite. We haven't seen any such code in the wild yet. The core team discussed that specific source compatibility concern and decided that the benefits of this change vastly outweigh the cost of such breakage; if we see more widespread breakage, we'll consider additional mitigations in the compiler (e.g, ignoring redundant conformances in such cases)

natecook1000 commented:
Thanks @DougGregor, glad to know it's been discussed!
